src/V3SchedTiming.cpp

```diff
@@ -0,0 +1,338 @@
+            auto* const activep = new AstActive{flp, "_timing", sensesp};
+            activep->addStmtsp(resumep);
+            m_lbs.emplace_back(m_scopeTopp, activep);
+        }
+
+        // VISITORS
+        virtual void visit(AstCAwait* nodep) override {
+            if (AstSenTree* const sensesp = nodep->sensesp()) {
+                if (!sensesp->user1SetOnce()) createResumeActive(nodep);
+                nodep->clearSensesp();  // Clear as these sentrees will get deleted later
+                if (m_inProcess) m_processDomains.insert(sensesp);
+            }
+        }
+        virtual void visit(AstNodeProcedure* const nodep) override {
+            m_inProcess = true;
+            iterateChildren(nodep);
+            // Mark vars as written by suspendable, add external domains
+            nodep->foreach<AstNodeVarRef>([&](AstNodeVarRef* const refp) {
+                if (refp->access().isWriteOrRW()) {
+                    const AstVarScope* const vscp = refp->varScopep();
+                    m_externalDomains[vscp].insert(m_processDomains.begin(),
+                                                   m_processDomains.end());
+                    vscp->varp()->setWrittenBySuspendable();
+                }
+            });
+            m_processDomains.clear();
+            m_inProcess = false;
+        }
+
+        //--------------------
+        virtual void visit(AstNodeMath*) override {}  // Accelerate
+        virtual void visit(AstNode* nodep) override { iterateChildren(nodep); }
+
+    public:
+        // CONSTRUCTORS
+        explicit AwaitVisitor(AstNetlist* nodep, LogicByScope& lbs,
+                              std::map<const AstVarScope*, std::set<AstSenTree*>>& externalDomains)
+            : m_scopeTopp{nodep->topScopep()->scopep()}
+            , m_lbs{lbs}
+            , m_externalDomains{externalDomains} {
+            iterate(nodep);
+        }
+        virtual ~AwaitVisitor() override = default;
+    };
+    TimingKit kit;
+    AwaitVisitor{netlistp, kit.m_lbs, kit.m_externalDomains};
+    return kit;
+}
+
+//============================================================================
+// Visits all forks and transforms their sub-statements into separate functions.
+
+void transformForks(AstNetlist* const netlistp) {
+    if (!v3Global.usesTiming()) return;
+    // Transform all forked processes into functions
+    class ForkVisitor final : public VNVisitor {
+    private:
+        // NODE STATE
+        //  AstVar::user1()  -> bool.  Set true if the variable was declared before the current
+        //                             fork.
+        const VNUser1InUse m_inuser1;
+
+        // STATE
+        bool m_inClass = false;  // Are we in a class?
+        AstFork* m_forkp = nullptr;  // Current fork
+        AstCFunc* m_funcp = nullptr;  // Current function
+
+        // METHODS
+        // Remap local vars referenced the function our statement
```

Comment:
```suggestion
        // Remap local vars referenced the function our statement
        // TODO: We should only pass variables to the fork that are
        // live in the fork body, but for that we need a proper data
        // flow analysis framework which we don't have at the moment.
```

This currently generates fork functions that are of the form:

```
fork(syms, whatnot) {
 whatnot = syms->whatnot + 1;
 co_await
 sym->other = whatnot;
}
```
`whatnot` is ought to be local but I think you need proper liveness data for that, but correct me if I'm wrong.

Reply:
You're correct.

Reply:
@kbieganski File issue for this as well please